### PR TITLE
[Dynamo][9/N] Make SkipFilesVariable wrap functions only

### PR DESCRIPTION
**Pull request by yanboliang**
Make ```SkipFilesVariable``` only handle function type, and route skipped classes to ```UserDefinedClassVariable```. The reasons behind this are:
* We'd like to remove ```is_allowed```, so the allowed/disallowed torch classes should have a proper place to handle. We can put them in either ```SkipFilesVariable``` and ```UserDefinedClassVariable``` under the current architecture, but it's  confusing to have two places do one thing.
   - Going forward, let's make ```SkipFilesVariable``` only handle functions, and probably I'll rename it to ```SkippedFunctionVariable``` in the following PRs.
   - Let's do dispatch by value's type, all torch classes stuff would go to ```UserDefinedClassVariable``` in the next PR.
* We'd merge in_graph/skip/inline trace decision into the same API ```trace_rule.lookup```, so probably we have to limit the input to only function for better organizing ```VariableBuilder._wrap``` logics.
   - Next step, I'll merge ```skipfiles.check``` into ```trace_rules.lookup```, and do the skipfile check before wrapping them into correct variable tracker.
   - Though the ```TorchCtxManagerClassVariable``` is decided by ```trace_rules.lookup```, I'll refactor it out in the following PRs.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/utils.py [hunk 1]
```diff
@@ -517,7 +517,6 @@ def is_function(value):
         value,
         (
             types.FunctionType,
-            types.MethodType,
```
Comment (yanboliang) [inline torch/_dynamo/utils.py:520]: I think this is typo in my previous PR, general method (non-nn.module) should be wrapped as ```UserDefinedObjectVariable``` as the binding object matters.

File: torch/_dynamo/variables/user_defined.py [hunk 2]
```diff
@@ -110,9 +128,44 @@ def call_function(
     ) -> "VariableTracker":
         from ..side_effects import SideEffects
         from .builder import SourcelessBuilder
+        from .builtin import BuiltinVariable
 
         if self.value is contextlib.nullcontext:
             return NullContextVariable()
+        elif self.value is collections.OrderedDict:
```
Comment (yanboliang) [inline torch/_dynamo/variables/user_defined.py:135]: For a lot of logics around ```collection.OrderedDict``` in UDC and UDO, we should do further refactor. I think there are two major things we need to figure out:
* Supporting arbitrary class as dict key - The challenging is how to generate global reference for these classes during installing guard. I was trying to fix this in this PR, but I found it's non-trivial, so I'll leave it as a separate PR. Or probably we only support ```int/float/str``` as key, arbitrary classes are not very common cases. If we do this, then many logics are not necessary, as they will be handled in ```ConstDictVariable```.
* Custom dict support - probably we need to consolidate these code.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**